internal/overlord/servstate/handlers.go

```diff
@@ -51,183 +54,528 @@ var (
+			if s.backoffIndex >= len(backoffDurations) {
+				// No more backoffs, transition to stopped state.
+				logger.Noticef("Service %q %s action is %q: no more backoffs", s.config.Name, onType, action)
+				s.transition(stateStopped)
+				return nil
 			}
-			releasePlan()
+			duration := backoffDurations[s.backoffIndex]
+			logger.Noticef("Service %q %s action is %q, waiting %s before restart (backoff %d/%d)",
+				s.config.Name, onType, action, duration, s.backoffIndex+1, len(backoffDurations))
+			s.backoffIndex++
+			s.transition(stateBackoffWait)
+			time.AfterFunc(duration, func() { logError(s.backoffWaitElapsed()) })
+
+		default:
+			return fmt.Errorf("internal error: unexpected action %q", action)
 		}
-		return fmt.Errorf("cannot start service: exited quickly with code %d", cmd.ProcessState.ExitCode())
+
+	case stateTerminating, stateKilling:
+		s.stopped <- nil
+		s.transition(stateStopped)
+
+	default:
+		return fmt.Errorf("exited invalid in state %q", s.state)
 	}
-	// unreachable
+	return nil
 }
 
-func (m *ServiceManager) doStop(task *state.Task, tomb *tomb.Tomb) error {
-	releasePlan, err := m.acquirePlan()
-	if err != nil {
-		return err
+// exitCode returns the exit code of the given command, or 128+signal if it
+// exited via a signal.
+func exitCode(cmd *exec.Cmd) int {
+	status, ok := cmd.ProcessState.Sys().(syscall.WaitStatus)
+	if ok {
+		if status.Signaled() {
+			return 128 + int(status.Signal())
+		}
+		return status.ExitStatus()
 	}
-	defer releasePlan()
+	return cmd.ProcessState.ExitCode()
+}
 
-	m.state.Lock()
-	req, err := TaskServiceRequest(task)
-	m.state.Unlock()
-	if err != nil {
-		return err
+// getAction returns the correct action to perform from the plan and whether
+// or not the service exited with a success exit code (0).
+func getAction(config *plan.Service, success bool) (action plan.ServiceAction, onType string) {
+	switch {
+	case !success && config.OnFailure != "":
+		return config.OnFailure, "on-failure"
+	case success && config.OnSuccess != "":
+		return config.OnSuccess, "on-success"
+	default:
+		onExit := config.OnExit
+		if onExit == "" {
+			onExit = plan.ActionRestart // default for "on-exit"
+		}
+		return onExit, "on-exit"
 	}
+}
 
-	active, ok := m.services[req.Name]
-	if !ok {
-		// Already stopped
-		m.state.Lock()
-		task.Logf("Service %q already stopped.", req.Name)
-		m.state.Unlock()
-		return nil
+// sendSignal sends the given signal to a running service.
+func (s *service) sendSignal(signal string) error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	switch s.state {
+	case stateStarting, stateRunning:
+		err := syscall.Kill(-s.cmd.Process.Pid, unix.SignalNum(signal))
+		if err != nil {
+			return err
+		}
+
+	case stateBackoffWait, stateTerminating, stateKilling, stateStopped:
+		return fmt.Errorf("cannot send signal while service is stopped or stopping")
+
+	default:
+		return fmt.Errorf("sendSignal invalid in state %q", s.state)
```

Comment:
Missing "internal error: " prefix? Would also be nice to avoid referring to the method name, as code gets refactored and this message would be left behind. Maybe: `"internal error: attempted to signal service in state %q"

Reply:
This isn't actually an internal error, but is bubbled up to the user in `ServiceManager.SendSignal()`. However, good point about the method name/refactoring, so I've removed the `sendSignal` name and now have: `"sending signal invalid in state %q"`.